Fill out `docs/Testing.md` (#2196)
infinitered/ignite: docs/Testing.md

- Given - some precondition
- When - some action executed by the function that you’re testing
- Then - the expected outcome

This is also known as AAA (Arrange, Act, Assert).

You can read more about how to best practices for creating tests in the [Structuring Tests](https://reactnative.dev/docs/testing-overview#structuring-tests) section of the React Native docs.

### Writing Tests

To write your own tests, create a `.test.ts` file within the `app` or `test` directory.

Then run `yarn test` to run all unit tests using Jest.

When writing tests, you can also run Jest in watch mode by running `yarn test:watch`. This will start a long running Jest process, that re-runs your tests on save in your editor. This is useful for iterating on values and getting quick feedback about whether your changes were successful or not.

### When to write Unit Tests

Some developers believe that every line of code should have a co-responding test. However, Infinite Red believes that some unit tests provide more values than others. Here are a few situations where we have found unit tests to be helpful in client projects:

> - _complicated regex_: writing a series of valid and invalid inputs can help track down sneaky bugs

Reviewer: Totally optional because everything reads great, but these 3 cases have a mix of capitalization and punctuation. Can make them consistent if you want.

Author: I'm pretty sure I got it consistent now, but if you could take another look, I would appreciate it

In addition to regular Javascript libraries, you can also mock out native modules in React Native. Using the following syntax

```
jest.mock('react-native-video', () => 'Video');
```

The first argument of `jest.mock` is the name of the module want to mock, but you can also pass a second argument to provide a function that returns the module.

In this example, this will return a default export.

You can read more in the [Testing React Native](https://jestjs.io/docs/tutorial-react-native#mock-native-modules-using-jestmock) section of the Jest docs.

## Resources

### Libraries

There are a variety of testing libraries available in React Native that you may find useful to add to your Ignite app

- [React Native Testing Library](https://callstack.github.io/react-native-testing-library/) - port of @testing-library/react for React Native. Useful for unit testing components.
- [Appium](https://appium.io/) - alternative for integration testing to Detox.

### Relevant Reading

- [React Native docs on Testing](https://reactnative.dev/docs/testing-overview)
- [Testing React Native Apps with Jest](https://jestjs.io/docs/tutorial-react-native)
- [Detox docs](https://wix.github.io/Detox/)

> - [Ken C. Dodds articles on Testing](https://kentcdodds.com/blog?q=test)

Reviewer: Ken => Kent

### Best Practices

> When writing a test, do your best to make sure that your tests > include the following information:
>
> - Given - some precondition
> - When - some action executed by the function that you’re testing
> - Then - the expected outcome
>   This is also known as AAA (Arrange, Act, Assert).

-- <cite>[Structuring Tests in React Native Docs](https://reactnative.dev/docs/testing-overview#structuring-tests)</cite>

You can read more about how to best practices for creating tests in the [Structuring Tests](https://reactnative.dev/docs/testing-overview#structuring-tests) section of the React Native docs.

### Writing Tests

To write your own tests, create a `.test.ts` file within the `app` or `test` directory.

Then run `yarn test` to run all unit tests using Jest.

When writing tests, you can also run Jest in watch mode by running `yarn test:watch`. This will start a long running Jest process, that re-runs your tests on save in your editor. This is useful for iterating on values and getting quick feedback about whether your changes were successful or not.

### When to write Unit Tests

> Some developers believe that every line of code should have a co-responding test. However, Infinite Red believes that some unit tests provide more values than others. Here are a few situations where we have found unit tests to be helpful in client projects:

Reviewer 2: Current phrasing focuses a little much on the contrasting of opinions, I think; I think we can just focus on what we recommend. Here's an example alternative:

> The most important question to ask when writing tests is "what code should be unit tested?" Not every line of code will benefit from a unit test. Typically, you'll want a unit test when you have code that can be run without external dependencies (like an API) that has some non-trivial logic.

I don't think that's the best wording, but it's an example :)

> Some developers believe that every line of code should have a co-responding test. However, Infinite Red believes that some unit tests provide more values than others. Here are a few situations where we have found unit tests to be helpful in client projects:

Reviewer 2: ```suggestion
Some developers believe that every line of code should have a corresponding test. However, Infinite Red believes that some unit tests provide more values than others. Here are a few situations where we have found unit tests to be helpful in client projects:
```

> - **complicated regex's**: for many developers, regex's only make sense as you are writing them. Testing with a series of valid and invalid inputs can help ensure that they work as intended for future developers.

Reviewer 2: ```suggestion
- **complicated regexes**: for many developers, regex's only make sense as you are writing them. Testing with a series of valid and invalid inputs can help ensure that they work as intended for future developers.
```

- **nested if/else statements**: if you rely heavily on a function with a lot of if/else statements, it can be helpful to have tests to make sure that you can visit each condition. Often times when we have more than a handful of conditionals, it can become impossible to visit all of them without realizing it.

> - **validation functions**: often times we may write functions like `isJson()` to validate that a value is a specific shape. If critical parts of our code rely on the correctness of this function, it can be useful to add tests to make sure that it fails on invalid inputs and passes on valid inputs.

Reviewer 2: Think this can be more concise, one idea:
```suggestion
- **validation functions**: often times we may write functions like `isJson()` to validate that a value is a specific shape. If critical parts of our code rely on the correctness of this function, we want to test that code!
```

## Mocking

However, often our code is not entirely written as pure functions. Most apps have side effects like making network requests, calling native modules, or reaching for global objects.

Relying on external code can make our code difficult and slow to test.

> One solution to making our tests easier to write is by utilizing mocks.

Reviewer 2: We say "one solution" but then don't enumerate anything else. I think it's important to mention somewhere here that if code's functionality relies on external systems, it may be more appropriate to test that code with E2E tests, rather than using mocking.

Author: I've updated that phrasing here: https://github.com/infinitered/ignite/pull/2196/commits/df5220851f804656deaa273253f6640b537ac360
